File: proyecto3/chat_front2/lib/controladores/card_controller.dart
import 'dart:convert';
import 'package:http/http.dart' as http;

import '../modelos/card.dart';

class CardController {
  final String baseUrl = 'http://localhost:8080/api/cards';

  // Obtener las tarjetas de un usuario
  Future<List<CreditCard>> fetchCards(String userId) async {
    final response = await http.get(Uri.parse('$baseUrl/$userId'));

    // Agregar un mensaje para saber si la consulta fue exitosa
    if (response.statusCode == 200) {
      print('Consulta exitosa');
      print('Respuesta del servidor: ${response.body}');

      final List<dynamic> data = json.decode(response.body);
      return data.map((card) => CreditCard.desdeJson(card)).toList();
    } else {
      print('Error en la consulta. Código de estado: ${response.statusCode}');
      throw Exception('Error al obtener las tarjetas');
    }
  }
  // Obtener una tarjeta por ID
  Future<CreditCard> fetchCardById(String cardId) async {
    final response = await http.get(Uri.parse('$baseUrl/card/$cardId'));

    if (response.statusCode == 200) {
      print('Tarjeta obtenida exitosamente');
      return CreditCard.desdeJson(json.decode(response.body));
    } else {
      print('Error al obtener la tarjeta. Código de estado: ${response.statusCode}');
      throw Exception('Error al obtener la tarjeta');
    }
  }

  // Congelar una tarjeta
  Future<void> freezeCard(String cardId) async {
    final response = await http.post(Uri.parse('$baseUrl/freeze/$cardId'));
    if (response.statusCode != 204) { // Cambia el código de estado esperado a 204
      print('Error al congelar la tarjeta. Código de estado: ${response.statusCode}');
      throw Exception('Error al congelar la tarjeta');
    } else {
      print('Tarjeta congelada exitosamente');
    }
  }


  // Eliminar una tarjeta
  Future<void> deleteCard(String cardId) async {
    final response = await http.delete(Uri.parse('$baseUrl/$cardId'));
    if (response.statusCode != 204) {
      print('Error al eliminar la tarjeta. Código de estado: ${response.statusCode}');
      throw Exception('Error al eliminar la tarjeta ${response.statusCode}');
    } else {
      print('Tarjeta eliminada exitosamente');
    }
  }

  // Agregar una tarjeta
  Future<void> addCard(String userId, CreditCard card) async {
    final response = await http.post(
      Uri.parse('$baseUrl/$userId'),
      headers: {'Content-Type': 'application/json'},
      body: json.encode(card.aJson()), // Suponiendo que el modelo tiene un método aJson()
    );

    if (response.statusCode != 201) {
      print('Error al agregar la tarjeta. Código de estado: ${response.statusCode}');
      throw Exception('Error al agregar la tarjeta');
    } else {
      print('Tarjeta agregada exitosamente');
    }
  }
}

File: proyecto3/banca-movil/src/main/java/com/espe/banca_movil/controller/CardController.java
package com.espe.banca_movil.controller;

import com.espe.banca_movil.model.Card;
import com.espe.banca_movil.service.CardService;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

@RestController
@RequestMapping("/api/cards")
public class CardController {

    @Autowired
    private CardService cardService;

    // Obtener todas las tarjetas de un usuario
    @GetMapping("/{userId}")
    public ResponseEntity<List<Card>> getCardsByUserId(@PathVariable Long userId) {
        return ResponseEntity.ok(cardService.getCardsByUserId(userId));
    }

    // Obtener una tarjeta por ID
    @GetMapping("/card/{id}")
    public ResponseEntity<Card> getCardById(@PathVariable Long id) {
        return ResponseEntity.ok(cardService.getCardById(id));
    }

    // Agregar una tarjeta
    @PostMapping("/{userId}")
    public ResponseEntity<Card> addCard(@PathVariable Long userId, @RequestBody Card card) {
        Card createdCard = cardService.addCard(userId, card);  // Pasamos el userId y la tarjeta
        return ResponseEntity.status(201).body(createdCard);  // Retornamos el código de estado 201 para creación
    }

    // Congelar tarjeta
    @PostMapping("/freeze/{id}")
    public ResponseEntity<Void> freezeCard(@PathVariable Long id) {
        cardService.freezeCard(id);
        return ResponseEntity.noContent().build();
    }

    // Eliminar tarjeta
    @DeleteMapping("/{id}")
    public ResponseEntity<Void> deleteCard(@PathVariable Long id) {
        cardService.deleteCard(id);
        return ResponseEntity.noContent().build();  // Código de respuesta para operación exitosa sin cuerpo
    }
}
